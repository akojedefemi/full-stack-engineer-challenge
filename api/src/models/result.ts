import { Entity, Column, PrimaryGeneratedColumn, CreateDateColumn, UpdateDateColumn } from 'typeorm';
import { Length, IsDate, IsNotEmpty, IsOptional } from 'class-validator';

export enum ResultStatus {
    QUEUED = "Queued",
    IN_PROGRESS = "In Progress",
    SUCCESS = "Success",
    FAILURE = "Failure",
}

@Entity()
export class Result {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({ type: 'enum', enum: ResultStatus, default: ResultStatus.QUEUED })
    status: string;

    @Column({ length: 80 })
    @Length(3, 80)
    @IsNotEmpty()
    repositoryName: string;

    @Column({ type: 'jsonb' })
    @IsNotEmpty()
    findings: object;

    @CreateDateColumn()
    queuedAt: Date;

    @Column({ nullable: true })
    @IsDate()
    @IsOptional()
    scanningAt: Date;

    @Column({ nullable: true })
    @IsDate()
    @IsOptional()
    finishedAt: Date;

    @UpdateDateColumn()
    updatedAt: Date;
}
export default Result;

// SCHEMA
const findingsExample = [{
    "type": "sast",
    "ruleId": "G402",
    "location": {
        "path": "connectors/apigateway.go",
        "positions": {
            "begin": {
                "line": 60
            }
        }
    },
    "metadata": {
        "description": "TLS InsecureSkipVerify set true.",
        "severity": "HIGH"
    }
}];
const exampleDate = new Date();

export const resultSchema = {
    id: { type: 'number', required: true, example: 1 },
    status: { type: 'string', required: false, example: 'Queued' },
    repositoryName: { type: 'string', required: true, example: 'my repository' },
    findings: { type: 'jsonb', required: true, example: findingsExample },
    queuedAt: { type: 'date', required: false, example: exampleDate },
    scanningAt: { type: 'date', required: false, example: exampleDate },
    finishedAt: { type: 'date', required: false, example: exampleDate },
    updatedAt: { type: 'date', required: false, example: exampleDate },
};