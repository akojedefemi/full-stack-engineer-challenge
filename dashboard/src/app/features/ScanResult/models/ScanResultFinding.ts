export interface ScanResultFinding {
  type: string;
  ruleId: string;
  location: Location;
  metadata: Metadata;
}

interface Location {
  path: string;
  positions: {
    begin: {
      line: number;
    }
  }
}

interface Metadata {
  severity: string;
  description: string;
}