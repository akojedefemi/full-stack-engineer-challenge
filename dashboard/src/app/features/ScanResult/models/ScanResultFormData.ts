export interface ScanResultFormData {
  repositoryName: string;
  findings: string;
}