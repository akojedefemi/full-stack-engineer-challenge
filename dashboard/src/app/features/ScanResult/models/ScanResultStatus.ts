export enum ScanResultStatus {
  QUEUED = "Queued",
  IN_PROGRESS = "In Progress",
  SUCCESS = "Success",
  FAILURE = "Failure",
}
