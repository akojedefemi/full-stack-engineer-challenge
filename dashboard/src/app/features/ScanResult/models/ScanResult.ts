import { ScanResultStatus } from "./ScanResultStatus";
import { ScanResultFinding } from "./ScanResultFinding";

export interface ScanResult {
  id: number;
  status: ScanResultStatus;
  repositoryName: string;
  findings: ScanResultFinding[];
  queuedAt: Date;
  scanningAt?: Date;
  finishedAt?: Date;
  updatedAt: Date;
}
