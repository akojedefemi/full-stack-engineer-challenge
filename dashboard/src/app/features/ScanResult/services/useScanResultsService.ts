import { useEffect, useState } from 'react';

import { ScanResult } from '../models/ScanResult';
import { Service } from '../../Service';
import { normalizeResult } from './normalizeResult';
import { config } from '../../../config';

const useScanResultsService = () => {
  const [result, setResult] = useState<Service<ScanResult[]>>({
    status: 'loading'
  });

  useEffect(() => {
    fetch(`${config.apiUrl}/api/v1/results`)
      .then(response => response.json())
      .then(response => response.map(normalizeResult))
      .then(response => setResult({ status: 'loaded', payload: response }))
      .catch(error => setResult({ status: 'error', error }));
  }, []);

  return result;
};

export default useScanResultsService;