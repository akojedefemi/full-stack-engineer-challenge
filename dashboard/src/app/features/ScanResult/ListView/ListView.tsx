import React from 'react';

import useScanResultsService from '../services/useScanResultsService';
import { ScanResult } from '../models/ScanResult';
import { Table } from 'antd';
import { ColumnProps } from 'antd/lib/table';
import { Link } from 'react-router-dom';
import { Routes } from '../../../Routes';

const renderTable = (data: ScanResult[]) => {

  const columns: ColumnProps<ScanResult>[] = [
    {
      title: 'Repository',
      dataIndex: 'repositoryName',
      key: 'repositoryName',
    },
    {
      title: 'Status',
      dataIndex: 'status',
      key: 'status',
    },
    {
      title: 'Last Change',
      dataIndex: 'updatedAt',
      key: 'updatedAt ',
      render: (date: Date) => date.toLocaleString(),
    },
    {
      title: 'View Findings',
      dataIndex: 'id',
      key: 'id ',
      render: (id: string) => <Link to={Routes.SCAN_RESULT_DETAILS.replace(':id', id)}>Findings</Link>,
    },
  ];

  return (
    <Table columns={columns} dataSource={data} rowKey="id" />
  );
}

const ListView: React.FC = () => {
  const service = useScanResultsService();

  return (
    <div>
      {service.status === 'loading' && <div>Loading...</div>}
      {service.status === 'loaded' && renderTable(service.payload)}
      {service.status === 'error' && (
        <div>Error, there was an network problem. Please try again later.</div>
      )}
    </div>
  )
};

export default ListView;
