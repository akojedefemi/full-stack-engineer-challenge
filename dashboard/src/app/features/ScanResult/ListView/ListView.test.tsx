import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

jest.mock('../services/useScanResultsService');

import ListView from './ListView';
import { ScanResultStatus } from '../models/ScanResultStatus';
import useScanResultsService from '../services/useScanResultsService';

const useScanResultsServiceMock = useScanResultsService as jest.Mock<ReturnType<typeof useScanResultsService>>;


const mockedScanResult = {
  id: 1,
  status: ScanResultStatus.QUEUED,
  repositoryName: 'someName',
  findings: [],
  queuedAt: new Date('2019-11-10T03:02:23.200Z'),
  scanningAt: undefined,
  finishedAt: undefined,
  updatedAt: new Date('2019-11-10T03:02:23.200Z'),
};

describe('ListView', () => {
  beforeEach(() => jest.resetAllMocks());

  it('renders loading state', () => {
    useScanResultsServiceMock.mockReturnValue({
      status: 'loading',
    });

    expect(renderer.create(<BrowserRouter><ListView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

  it('renders loaded state', () => {
    useScanResultsServiceMock.mockReturnValue({
      status: 'loaded',
      payload: [mockedScanResult]
    });

    expect(renderer.create(<BrowserRouter><ListView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

  it('renders error state', () => {
    useScanResultsServiceMock.mockReturnValue({
      status: 'error',
      error: new Error('some error'),
    });

    expect(renderer.create(<BrowserRouter><ListView /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

});
