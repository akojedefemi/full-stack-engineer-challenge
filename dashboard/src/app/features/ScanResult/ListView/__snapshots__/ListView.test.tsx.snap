// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders error state 1`] = `
<div>
  <div>
    Error, there was an network problem. Please try again later.
  </div>
</div>
`;

exports[`ListView renders loaded state 1`] = `
<div>
  <div
    className="ant-table-wrapper"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-default ant-table-scroll-position-left"
          style={Object {}}
        >
          <div
            className="ant-table-content"
          >
            <div
              className="ant-table-body"
              onScroll={[Function]}
              onWheel={[Function]}
              style={Object {}}
            >
              <table
                className=""
                style={Object {}}
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                </colgroup>
                <thead
                  className="ant-table-thead"
                >
                  <tr
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            Repository
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            Status
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            Last Change
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            View Findings
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key={1}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      someName
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      Queued
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      11/10/2019, 10:02:23 AM
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <a
                        href="/scan-result/1"
                        onClick={[Function]}
                      >
                        Findings
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          className="ant-pagination ant-table-pagination"
          style={Object {}}
          unselectable="unselectable"
        >
          <li
            aria-disabled={true}
            className="ant-pagination-disabled ant-pagination-prev"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Previous Page"
          >
            <a
              className="ant-pagination-item-link"
            >
              <i
                aria-label="icon: left"
                className="anticon anticon-left"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </i>
            </a>
          </li>
          <li
            className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex="0"
            title={1}
          >
            <a>
              1
            </a>
          </li>
          <li
            aria-disabled={true}
            className="ant-pagination-disabled ant-pagination-next"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Next Page"
          >
            <a
              className="ant-pagination-item-link"
            >
              <i
                aria-label="icon: right"
                className="anticon anticon-right"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                  />
                </svg>
              </i>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListView renders loading state 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
